### ISSUE TITLE
Allow for a named tuple for `initial_params`

### ISSUE THREAD

**Issue by DominiqueMakowski** [TuringLang/Turing.jl#2286]
My understanding is that currently `initial_params` must be a vector of the length of the parameters, but this can become a bit tricky for models with lots of parameters. 

Would it be possible to allow for passing, for instance, a named tuple (`initial_params=(μ=2.5, σ=1.0)`) to set initial values on desired parameters?

**Comment by Red-Portal** [comment]
I would also love to see named tuple generally supported across the Turing ecosystem, but that would probably need some work. It's also somewhat dependent on [this PR](https://github.com/JuliaStats/Distributions.jl/pull/1803)

**Comment by torfjelde** [comment]
> Would it be possible to allow for passing, for instance, a named tuple (initial_params=(μ=2.5, σ=1.0)) to set initial values on desired parameters?

Definitively possible! I made a quick PoC concept here: https://github.com/TuringLang/DynamicPPL.jl/pull/632

I'm somewhat limited in time these days, but maybe someone can complete it (shouldn't be much work).

> I would also love to see named tuple generally supported across the Turing ecosystem, but that would probably need some work. It's also somewhat dependent on https://github.com/JuliaStats/Distributions.jl/pull/1803

Supporting `initial_params::NamedTuple` would actually be quite a bit simpler than supporting `NamedTuple` across the entire TuringLang ecosystem:)

**Comment by Red-Portal** [comment]
@torfjelde We don't yet officially support passing `NamedTuple`s to a `LogDensityFunction` right?

**Comment by torfjelde** [comment]
> We don't yet officially support passing NamedTuples to a LogDensityFunction right?

No, but we don't need this to support `initial_params::NamedTuple`.

**Comment by DominiqueMakowski** [comment]
Another potential feature for https://github.com/TuringLang/DynamicPPL.jl/pull/632 would be to set all/some of these initial parameters as functions that are then applied the prior. For instance, `mean()` or `mode()` (I reckon this would be quite convenient to use, like `sample(..., init_params=mean()`, to take the mean of all the priors as starting params)

**Comment by torfjelde** [comment]
Though I agree that seems convenient, it's just a bit too much hassle to maintain as it really only saves a single call from the user perspective:
```julia
init_parmas=mean(rand(Vector, model))`
```
(though this is somewhat "new" / not well-documented tbh)

**Comment by torfjelde** [comment]
This has now been solved by https://github.com/TuringLang/DynamicPPL.jl/pull/632 I believe? @sunxd3 ?:)

**Comment by DominiqueMakowski** [comment]
Cheers for the work!

**Comment by DominiqueMakowski** [comment]
Small thing, but `mean(rand(Vector, model))` does not return the mean for each parameter (but just the mean of a random draw from all the parameters). Thus I'm not sure how to initialize using the mean of the  prior distribution of *each* parameter?

**Comment by torfjelde** [comment]
> Small thing, but mean(rand(Vector, model)) does not return the mean for each parameter (but just the mean of a random draw from all the parameters). Thus I'm not sure how to initialize using the mean of the prior distribution of each parameter?

Ah sorry, yes that's very true. I wrote that a bit too quickly trying to convey that you can take the mean of smaoples from the model to get it.

You can do:
```julia
chain_prior = sample(model, Prior(), 1000)
```
and then extract from this:)

**Comment by DominiqueMakowski** [comment]
Well yes but my point was about some syntactic sugar to conveniently (and efficiently) set the initial parameters to the prior means, and sampling from the priors first appears as overkill given that the distributions usually have analytically defined means.

Hence the `initial_params=:mean` proposition or ``initial_params=mean` (i.e., passing a function that would be applied to each prior distribution). 

Alternatively, is there a way to extract the prior distributions as a vector, for instance:

```jl
priors = get_priors(model)  
priors
```
```
(mu=Normal(0, 1), sigma=Gamma(2, 2))
```

One could then run:

`initial_params=mean.(priors)`
